File: packages/sdk/sdk-py/src/phaseo/jobs.py
"""Handwritten polling helpers shared by music, video, and batches."""
from __future__ import annotations

import math
import time
from threading import Event
from typing import Any, Callable, Literal
from typing_extensions import TypedDict, Unpack

JobKind = Literal["music", "video", "batch"]
JobResponse = dict[str, Any]


class JobHandle:
    def __init__(self, kind: JobKind, job_id: str, retrieve: Callable[[str], JobResponse],
                 initial: JobResponse | None = None, cancel: Callable[[str], JobResponse] | None = None):
        if not job_id.strip():
            raise ValueError("Job ID is required")
        self.kind, self.id, self._retrieve, self._initial, self._cancel = kind, job_id, retrieve, initial, cancel

    def to_dict(self) -> dict[str, str]:
        return {"kind": self.kind, "id": self.id}

    def result(self, **options: Unpack[JobWaitOptions]) -> JobResponse:
        response = wait_for_job(self.kind, self.id, self._retrieve, initial=self._initial, **options)
        if job_status(response) != "completed":
            raise JobFailedError(self.kind, response)
        return response

    def cancel(self) -> JobResponse:
        if self._cancel is None:
            raise NotImplementedError(f"Remote cancellation is not supported for {self.kind}")
        return self._cancel(self.id)

    def events(self, *, interval: float = 5, timeout: float = 1800):
        _validate_options({"interval": interval, "timeout": timeout})
        deadline = time.monotonic() + timeout
        last = self._initial
        initial = self._initial
        while True:
            if time.monotonic() >= deadline:
                raise JobTimeoutError(self.kind, self.id, last)
            last = initial if initial is not None else self._retrieve(self.id)
            initial = None
            if time.monotonic() >= deadline:
                raise JobTimeoutError(self.kind, self.id, last)
            yield last
            if job_status(last) in ("completed", "failed", "cancelled", "expired"):
                return
            time.sleep(min(max(.25, interval), max(0, deadline - time.monotonic())))


class JobWaitOptions(TypedDict, total=False):
    interval: float
    timeout: float
    on_poll: Callable[[JobResponse], None]
    cancel_event: Event


class JobTimeoutError(TimeoutError):
    def __init__(self, kind: JobKind, job_id: str, last_response: JobResponse | None):
        super().__init__(f"Timed out waiting for {kind} {job_id}")
        self.kind = kind
        self.job_id = job_id
        self.last_response = last_response


class JobCancelledError(Exception):
    """The local wait was cancelled; the remote job was not cancelled."""

    def __init__(self, kind: JobKind, job_id: str, last_response: JobResponse | None):
        super().__init__(f"Stopped waiting for {kind} {job_id}")
        self.kind = kind
        self.job_id = job_id
        self.last_response = last_response


class JobFailedError(Exception):
    def __init__(self, kind: JobKind, response: JobResponse):
        super().__init__(f"{kind} {response.get('id', '')} finished with status {job_status(response)}")
        self.kind = kind
        self.job_id = response.get("id")
        self.response = response


def job_status(job: JobResponse) -> str:
    status = str(job.get("status") or job.get("lifecycle_status") or "").strip().lower()
    return "cancelled" if status == "canceled" else status


def _validate_options(options: JobWaitOptions) -> None:
    for name in ("interval", "timeout"):
        value = options.get(name)
        if value is not None and (isinstance(value, bool) or not isinstance(value, (int, float)) or not math.isfinite(value) or value <= 0):
            raise ValueError(f"{name} must be a finite positive number")


def wait_for_job(
    kind: JobKind,
    job_id: str,
    retrieve: Callable[[str], JobResponse],
    *,
    initial: JobResponse | None = None,
    **options: Unpack[JobWaitOptions],
) -> JobResponse:
    """Return a terminal response, including failures. Check deadlines between HTTP calls."""
    _validate_options(options)
    job_id = job_id.strip()
    if not job_id:
        raise ValueError(f"{kind} ID is required")
    interval = max(0.25, options.get("interval", 5.0))
    deadline = time.monotonic() + options.get("timeout", 1800.0)
    cancel_event = options.get("cancel_event")
    last = initial

    def check_stopped() -> None:
        if cancel_event is not None and cancel_event.is_set():
            raise JobCancelledError(kind, job_id, last)
        if time.monotonic() >= deadline:
            raise JobTimeoutError(kind, job_id, last)

    while True:
        check_stopped()
        last = initial if initial is not None else retrieve(job_id)
        initial = None
        check_stopped()
        on_poll = options.get("on_poll")
        if on_poll is not None:
            on_poll(last)
        check_stopped()
        if job_status(last) in ("completed", "failed", "cancelled", "expired"):
            return last
        delay = min(interval, max(0, deadline - time.monotonic()))
        if cancel_event is not None:
            cancel_event.wait(delay)
        else:
            time.sleep(delay)


def create_and_wait_for_job(
    kind: JobKind,
    create: Callable[[], JobResponse],
    retrieve: Callable[[str], JobResponse],
    **options: Unpack[JobWaitOptions],
) -> JobResponse:
    """Submit once. The waiting deadline starts after the submission returns."""
    _validate_options(options)
    cancel_event = options.get("cancel_event")
    if cancel_event is not None and cancel_event.is_set():
        raise JobCancelledError(kind, "", None)
    initial = create()
    result = wait_for_job(kind, str(initial.get("id") or ""), retrieve, initial=initial, **options)
    if job_status(result) != "completed":
        raise JobFailedError(kind, result)
    return result

File: packages/sdk/sdk-ts/src/jobs.ts
import type { BatchResponse, MusicGenerateResponse, VideoStatusResponse } from "./index.js";

export type JobKind = "music" | "video" | "batch";
type JobResponses = { music: MusicGenerateResponse; video: VideoStatusResponse; batch: BatchResponse };
export type JobSnapshot = { id?: string; status?: string; lifecycle_status?: string };
export type JobWaitOptions<T> = {
  intervalMs?: number;
  timeoutMs?: number;
  signal?: AbortSignal;
  onPoll?: (job: T) => void | Promise<void>;
};

export class JobTimeoutError extends Error {
  constructor(readonly kind: JobKind, readonly jobId: string, readonly lastResponse?: JobSnapshot) {
    super(`Timed out waiting for ${kind} ${jobId}`);
    this.name = "JobTimeoutError";
  }
}

export class JobFailedError<K extends JobKind = JobKind> extends Error {
  constructor(readonly kind: K, readonly response: JobResponses[K]) {
    super(`${kind} ${response.id ?? ""} finished with status ${jobStatus(response)}`);
    this.name = "JobFailedError";
  }
  get jobId(): string | undefined { return this.response.id; }
  isKind<T extends JobKind>(kind: T): this is this & JobFailedError<T> {
    return (this.kind as JobKind) === kind;
  }
}

export class JobCancelledError extends Error {
  constructor(readonly kind: JobKind, readonly jobId: string, readonly lastResponse?: JobSnapshot, cause?: unknown) {
    super(`Stopped waiting for ${kind} ${jobId}`, { cause });
    this.name = "JobCancelledError";
  }
}

export function jobStatus(job: JobSnapshot): string {
  const status = (job.status || job.lifecycle_status || "").trim().toLowerCase();
  return status === "canceled" ? "cancelled" : status;
}

const TERMINAL = ["completed", "failed", "cancelled", "expired"];

export function validateWaitOptions<T>(options: JobWaitOptions<T>): void {
  for (const [name, value] of [["intervalMs", options.intervalMs], ["timeoutMs", options.timeoutMs]] as const) {
    if (value !== undefined && (!Number.isFinite(value) || value <= 0)) {
      throw new RangeError(`${name} must be a finite positive number`);
    }
  }
}

/** Waits on an existing job. Terminal failures are returned for inspection. */
export async function waitForJob<T extends JobSnapshot>(
  kind: JobKind,
  id: string,
  retrieve: (id: string, signal?: AbortSignal) => Promise<T>,
  options: JobWaitOptions<T> = {},
  initial?: T,
  terminalStatuses: readonly string[] = TERMINAL,
): Promise<T> {
  validateWaitOptions(options);
  const jobId = id.trim();
  if (!jobId) throw new Error(`${kind} ID is required`);
  if (options.signal?.aborted) throw new JobCancelledError(kind, jobId, initial, options.signal.reason);
  if (!terminalStatuses.length) throw new Error("At least one terminal status is required");
  const intervalMs = Math.max(250, options.intervalMs ?? 5_000);
  const timeoutMs = options.timeoutMs ?? 30 * 60_000;
  let last = initial;
  const deadline = performance.now() + timeoutMs;
  let timer: ReturnType<typeof setTimeout> | undefined;
  let abort: (() => void) | undefined;
  const controller = new AbortController();
  const stopped = new Promise<never>((_, reject) => {
    timer = setTimeout(() => {
      const error = new JobTimeoutError(kind, jobId, last);
      controller.abort(error);
      reject(error);
    }, timeoutMs);
    abort = () => {
      const error = new JobCancelledError(kind, jobId, last, options.signal?.reason);
      controller.abort(error);
      reject(error);
    };
    options.signal?.addEventListener("abort", abort, { once: true });
  });
  // A user callback can abort and then throw before the next race is attached.
  void stopped.catch(() => {});
  let sleepTimer: ReturnType<typeof setTimeout> | undefined;
  try {
    while (true) {
      if (options.signal?.aborted) throw new JobCancelledError(kind, jobId, last, options.signal.reason);
      if (performance.now() >= deadline) throw new JobTimeoutError(kind, jobId, last);
      last = initial ?? await Promise.race([retrieve(jobId, controller.signal), stopped]);
      initial = undefined;
      if (options.onPoll) await Promise.race([Promise.resolve(options.onPoll(last)), stopped]);
      if (options.signal?.aborted) throw new JobCancelledError(kind, jobId, last, options.signal.reason);
      if (performance.now() >= deadline) throw new JobTimeoutError(kind, jobId, last);
      if (terminalStatuses.includes(jobStatus(last))) return last;
      await Promise.race([
        new Promise<void>((resolve) => { sleepTimer = setTimeout(resolve, intervalMs); }),
        stopped,
      ]);
    }
  } finally {
    clearTimeout(timer);
    clearTimeout(sleepTimer);
    if (abort) options.signal?.removeEventListener("abort", abort);
  }
}

/** Submit exactly once; timeoutMs applies to waiting after submission returns. */
export async function createAndWaitForJob<K extends JobKind>(
  kind: K,
  create: () => Promise<JobResponses[K]>,
  retrieve: (id: string, signal?: AbortSignal) => Promise<JobResponses[K]>,
  options: JobWaitOptions<JobResponses[K]> = {},
): Promise<JobResponses[K]> {
  validateWaitOptions(options);
  options.signal?.throwIfAborted();
  const initial = await create();
  const result = await waitForJob(kind, initial.id ?? "", retrieve, options, initial);
  if (jobStatus(result) !== "completed") throw new JobFailedError(kind, result);
  return result;
}
